src/boot/efi/cpio.c

```diff
@@ -336,9 +337,24 @@ EFI_STATUS pack_cpio(
                 return EFI_SUCCESS;
         }
 
-        root = LibOpenRoot(loaded_image->DeviceHandle);
-        if (!root)
-                return log_error_status_stall(EFI_LOAD_ERROR, L"Unable to open root directory.");
+        err = uefi_call_wrapper(BS->HandleProtocol, 3, loaded_image->DeviceHandle,
+                                &FileSystemProtocol, (VOID*)&volume);
+        /* It will be unsupported if not vfat. This can happen if the efi was
+         * loaded by a bootloader with support for non-vfat partitions, like
+         * grub. Do not stall in that case.
+         */
+        if (err == EFI_UNSUPPORTED)
+                return err;
```

Comment:
So my understanding is that grub simply doesn't implement a file system driver at all, but just has its own loader, that doesn't implement the `FileSystemProtocol`?

if that's the case, then the commit msg/comments should be changed, to make this clear, the hint regrding ext4 is highly misleading, because it has little to do with the choice of backing fs but everything with the set of APIs Grub decided to implement on the backing handle

Reply:
Right, grub implements a few filesystems itself instead of relying on UEFI to load files. I will reformat the commit message to make this more clear.